src/smoosh/src/smoosh_channel.erl

```diff
@@ -194,21 +215,128 @@ handle_info(check_window, State0) ->
         end,
     erlang:send_after(60 * 1000, self(), check_window),
     {noreply, FinalState};
+handle_info(persist_queue, State0) ->
+    #state{waiting = Queue} = State0,
+    write_state_to_file(State0),
+    smoosh_priority_queue:write_to_file(Queue),
+    Checkpoint =
+        config:get_integer(
+            "smoosh", "state_checkpoint_interval_in_sec", ?DEFAULT_CHECKPOINT_INTERVAL_IN_SEC
+        ) * 1000,
+    erlang:send_after(Checkpoint, self(), persist_queue),
+    {noreply, State0};
 handle_info(pause, State0) ->
-    {ok, State} = code_change(nil, State0, nil),
+    State = maybe_open_queue(State0),
     {noreply, State#state{paused = true}};
 handle_info(unpause, State0) ->
-    {ok, State} = code_change(nil, State0, nil),
+    State = maybe_open_queue(State0),
     {noreply, maybe_start_compaction(State#state{paused = false})}.
 
-terminate(_Reason, _State) ->
+terminate(_Reason, #state{name = Name, waiting = Q}) ->
+    file:delete(active_file_name(Name)),
+    file:delete(starting_file_name(Name)),
+    if
+        Q =/= nil ->
+            smoosh_priority_queue:close(Q);
+        true ->
+            nil
+    end,
     ok.
 
-code_change(_OldVsn, #state{} = State, _Extra) ->
-    {ok, State}.
+maybe_recover_state(#state{name = Name} = State) ->
+    Active = recover(active_file_name(Name)),
+    Starting = recover(starting_file_name(Name)),
+    DatabaseDir = config:get("couchdb", "database_dir"),
+    ViewDir = config:get("couchdb", "view_index_dir"),
+    Active1 = get_matching_compact_files(DatabaseDir, Active),
+    Active2 = get_matching_compact_files(ViewDir, Active),
+    Active3 = Active1 ++ Active2,
+    State#state{active = Active3, starting = Starting}.
+
+get_matching_compact_files(Dir, Active) ->
+    MatchingFiles = filelib:fold_files(
+        Dir,
+        "^[a-zA-Z0-9_.-]*.compact$",
+        true,
+        (fun(FilePath, Acc) ->
+            FilePrefix = filename:rootname(FilePath, ".compact"),
+            case lists:keyfind(FilePrefix, 1, Active) of
```

Comment:
This is going to be slow. The `lists:keyfind/3` takes `O(n)` to get the value. We repeat the `lists:keyfind/3` for each file candidate. IRC `Active` list is a term of the following structure `[{FilePath, Pid}]`. In this case the following would work better (not tested pseudo-code):

```
maybe_recover_state(#state{name = Name} = State) ->
    Active = recover(active_file_name(Name)),
    ActiveMap = maps:from_list(Active),
    Active1 = get_matching_compact_files(DatabaseDir, ActiveMap),
    Active2 = get_matching_compact_files(ViewDir, ActiveMap),
    ....

get_matching_compact_files(Dir, ActiveMap) ->
    MatchingFilesMap = filelib:fold_files(
        Dir,
        "^[a-zA-Z0-9_.-]*.compact$",
        true,
        fun(FilePath, Acc) ->
            FilePrefix = filename:rootname(FilePath, ".compact"),
            maps:remove(FilePrefix, Acc)
        end, ActiveMap),
     maps:to_list(MatchingFilesMap).
```

With this version the delete from map is O(log(n)) and the n is getting smaller on each iteration. This function is probably the only one in the whole module which need to be scalable. Because on big clusters there might a lot of files in the state term and even more on the filesystem.

Reply:
Actually I had a confusion. After re-reading the https://github.ibm.com/cloudant/dbcore/issues/53 ticket I think we need to do the opposite. We need to remove compaction jobs from active list if the compact files are not present on disk.

There are two options here:

1. Iterate over Active and remove jobs if compact files are not present on disk. This wouldn't pick compaction jobs triggered between last checkpoint and node restart.  
2. Do not persist active jobs at all and always do a directory scan on smoosh start. Add all compact files found into Active. This would be slower since we would be scanning a directory with hundred thousands of files. However this PR does exactly this.

Reply:
I think we can make a compromise. We can choose to implement approach #1 in maybe_recover_state, and in addition we would implement an auxiliary function `smoosh:enqueue_recovered` with the intend to call it using remsh. 
We restrict `smoosh:enqueue_recovered` to dbs only and implement it via new helper `couch_engine:fold_compactions`. 

The assumption is here that it is ok to loose compaction jobs which are started less than ~mins ago. In cases where the disk space is really limited the operator might engage `smoosh:enqueue_recovered` manually. 

I think the `smoosh:enqueue_recovered` (and `couch_engine:fold_compactions`) could be a separate PR.

Reply:
In approach #1, the time complexity will be `O(n*m)` where `n` is the `length(Active)` and `m` is the number of files in the directory because for each entry in `Active` we will search all `m` files for a matching compact file. This is the same complexity as the current approach I believe.

Reply:
How about an approach where we for each compact file, we check the `ActiveMap` for the `FilePrefix` (which would be constant time). If it is not found, remove it. Otherwise, continue. Then, at the end convert the `ActiveMap` back to a list. Although, this still does not significantly reduce computation time.

Reply:
> In approach #1, the time complexity will be O(n*m) where n is the length(Active) and m is the number of files in the directory because for each entry in Active we will search all m files for a matching compact file. 

Expect we don't need to search. The object id is a shard name. We can add auxiliary function to `couch_db`.
One of (I like 2nd more, since it doesn't leak implementation details): 

- `couch_db:get_compact_files(DbName) -> {MetaFile, DataFile}.` 
- `couch_db:has_compact_files(DbName) -> true | false.` (or `couch_db:is_compacting/1`)

Due to additional abstraction via `couch_engine` it wouldn't be that trivial to do plumbing from `couch_db` to `couch_bt_engine_compactor`, where we would be defining the function, but nevertheless doable.

Reply:
In this approach, what is the need for the heavy abstraction? A less abstracted approach in `smoosh_channel.erl` based on your recommendations:
```
maybe_recover_state(#state{name = Name} = State) ->
    Active = recover(active_file_name(Name)),
    Starting = recover(starting_file_name(Name)),
    Active1 = lists:foldl(
        (fun({DbFilePath, Pid}, Acc) ->
            case is_compacting(DbFilePath) of
                true ->
                    [{DbFilePath, Pid} | Acc];
                false ->
                    Acc
            end
        end),
        [],
        Active
    ),
    State#state{active = Active1, starting = Starting}.

is_compacting(DbFilePath) ->
    DataFileExists = case file:read_file_info(DbFilePath ++ ".compact.data") of
        {ok, _} ->
            true;
        {error, _} ->
            false
    end,
    MetaFileExists = case file:read_file_info(DbFilePath ++ ".compact.meta") of
        {ok, _} ->
            true;
        {error, _} ->
            false
    end,
    DataFileExists or MetaFileExists.
```
This would have `O(n)` complexity.

Reply:
> In this approach, what is the need for the heavy abstraction? 


The filenames ending ".compact.data" and ".compact.meta" is an implementation detail. Using it directly violates abstraction boundaries. Currently the codebase is structured in such a way that storage engines are plugable. Using specifics of a btree engine breaks other engines. To be fair we don't have any official alternative engines at  the moment and might remove `couch_engine` abstraction in the future if community decides so.

Reply:
Okay, here is an attempt https://github.com/apache/couchdb/pull/3766/commits/7d7ebf5bed47de589a7f4607baa937a4c2826a3b.

Reply:
The `is_compacting` plumbing looks good 👍